### PR TITLE
ARROW-11563: [Rust] Support Cast(Utf8, TimeStamp(Nanoseconds, None))

### PR DESCRIPTION
**Pull request by ovr**
DataFushion already has this cast with its own implementation inside DF based on top of the chrono library. In this PR, I moved functions called string_to_timestamp_nanos and naive_datetime_to_timestamp (helpers) to Arrow's compute::kernels crate to make it possible to do Cast(Utf8, Timestamp(Nanoseconds, None)) and usage these helpers inside DF for to_timestamp function.

### CODE REVIEW

File: rust/arrow/src/compute/kernels/cast.rs [hunk 1]
```diff
@@ -1480,6 +1498,19 @@ mod tests {
         assert!(c.is_null(2));
     }
 
+    #[test]
+    fn test_cast_string_to_timestamp() {
+        let a = StringArray::from(vec![Some("2020-09-08T12:00:00+00:00"), None]);
```
Comment (alamb) [inline rust/arrow/src/compute/kernels/cast.rs:1503]: I recommend adding an error case here too -- like `Some("not a date")` to ensure / encode the make-a-null-if-invalid-parse behavior

File: rust/arrow/src/error.rs [hunk 2]
```diff
@@ -27,6 +27,7 @@ pub enum ArrowError {
     /// Returned when functionality is not yet available.
     NotYetImplemented(String),
     ExternalError(Box<dyn Error + Send + Sync>),
+    CastError(String),
```
Comment (alamb) [inline rust/arrow/src/error.rs:30]: Adding a new Arrow Error variant seems reasonable to me -- what do you think @jorgecarleitao  and @nevi-me and @andygrove ?

**Review by alamb — APPROVED** [review]
Nice work @ovr  -- I have an additional suggested test but this PR could also be merged in as is I think

File: rust/arrow/src/compute/kernels/cast.rs (hunk 1 above)
Comment (ovr) [reply rust/arrow/src/compute/kernels/cast.rs]: Sure, good catch 👍

**Comment by alamb** [comment]
I think this PR is ready to go after a rebase -- any thoughts @Dandandan  @jorgecarleitao or @andygrove ?

File: rust/arrow/src/error.rs (hunk 2 above)
Comment (nevi-me) [reply rust/arrow/src/error.rs:30]: I'm happy with the addition

**Review by nevi-me — APPROVED** [review]
LGTM

**Comment by ovr** [comment]
Just a notice: Rebased

**Comment by alamb** [comment]
Thanks again @ovr !

**Pull request closed without merging**